src/NBitcoin/ChainIndexer.cs

```diff
@@ -0,0 +1,248 @@
+    /// </summary>
+    public class ChainIndexer
+    {
+        /// <summary>Locks access to <see cref="blocksByHeight"/>, <see cref="blocksById"/>.</summary>
+        private readonly object lockObject = new object();
+
+        /// <remarks>This object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<int, ChainedHeader> blocksByHeight;
+
+        /// <remarks>This object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<uint256, ChainedHeader> blocksById;
+
+        public Network Network { get; }
+
+        /// <summary>
+        /// The tip of the best known validated chain.
+        /// </summary>
+        public virtual ChainedHeader Tip { get; private set; }
+
+        /// <summary>
+        /// The tip height of the best known validated chain.
+        /// </summary>
+        public int Height => this.Tip.Height;
+        public ChainedHeader Genesis => this.GetBlock(0);
+
+        public ChainIndexer()
+        {
+            this.blocksByHeight = new Dictionary<int, ChainedHeader>();
+            this.blocksById = new Dictionary<uint256, ChainedHeader>();
+        }
+
+        public ChainIndexer(Network network) : this()
+        {   
+            this.Network = network;
+
+            this.Initialize(new ChainedHeader(network.GetGenesis().Header, network.GetGenesis().GetHash(), 0));
+        }
+
+        public ChainIndexer(Network network, ChainedHeader chainedHeader) : this()
+        {
+            this.Network = network;
+
+            this.Initialize(chainedHeader);
+        }
+
+        public void Initialize(ChainedHeader chainedHeader)
+        {
+            lock (this.lockObject)
+            {
+                this.blocksById.Clear();
+                this.blocksByHeight.Clear();
+
+                ChainedHeader iterator = chainedHeader;
+
+                while (iterator != null)
+                {
+                    this.blocksById.Add(iterator.HashBlock, iterator);
+                    this.blocksByHeight.Add(iterator.Height, iterator);
+
+                    if (chainedHeader.Height == 0)
+                    {
+                        if (chainedHeader.HashBlock != iterator.HashBlock)
+                            throw new InvalidOperationException("Wrong network");
+                    }
+
+                    iterator = iterator.Previous;
+                }
+
+                this.Tip = chainedHeader;
+            }
+        }
+        
+        /// <summary>
+        /// Returns the first chained block header that exists in the chain from the list of block hashes.
+        /// </summary>
+        /// <param name="hashes">Hash to search for.</param>
+        /// <returns>First found chained block header or <c>null</c> if not found.</returns>
+        public ChainedHeader FindFork(IEnumerable<uint256> hashes)
+        {
+            if (hashes == null)
```

Comment:
Do we really need to throw exceptions here? I would perhaps get the caller to check this so that the method's only responsibility is to `FindFork`.

Reply:
This is the same functionality fro chain base.